cni/network/invoker_cns.go

```diff
@@ -127,97 +132,167 @@ func (invoker *CNSIPAMInvoker) Add(addConfig IPAMAddConfig) (IPAMAddResult, erro
 			hostSubnet:         response.PodIPInfo[i].HostPrimaryIPInfo.Subnet,
 			hostPrimaryIP:      response.PodIPInfo[i].HostPrimaryIPInfo.PrimaryIP,
 			hostGateway:        response.PodIPInfo[i].HostPrimaryIPInfo.Gateway,
-		}
-
-		// set the NC Primary IP in options
-		// SNATIPKey is not set for ipv6
-		if net.ParseIP(info.ncPrimaryIP).To4() != nil {
-			addConfig.options[network.SNATIPKey] = info.ncPrimaryIP
+			addressType:        response.PodIPInfo[i].AddressType,
+			macAddress:         response.PodIPInfo[i].MacAddress,
+			isDefaultInterface: response.PodIPInfo[i].IsDefaultInterface,
+			routes:             response.PodIPInfo[i].Routes,
 		}
 
 		log.Logger.Info("Received info for pod",
-			zap.Any("ipv4info", info),
-			zap.Any("podInfo", podInfo))
-		ip, ncIPNet, err := net.ParseCIDR(info.podIPAddress + "/" + fmt.Sprint(info.ncSubnetPrefix))
-		if ip == nil {
-			return IPAMAddResult{}, errors.Wrap(err, "Unable to parse IP from response: "+info.podIPAddress+" with err %w")
-		}
+			zap.Any("ipinfo", info),
+			zap.Any("podInfo", podInfo),
+			zap.String("component", "cni-invoker-cns"))
+
+		switch info.addressType {
+		case cns.Secondary:
+			ip, ipnet, err := response.PodIPInfo[i].PodIPConfig.GetIPNet()
+			if ip == nil {
+				return IPAMAddResult{}, errors.Wrap(err, "Unable to parse IP from response: "+info.podIPAddress+" with err %w")
+			}
 
-		ncgw := net.ParseIP(info.ncGatewayIPAddress)
-		if ncgw == nil {
-			// TODO: Remove v4overlay and dualstackoverlay options, after 'overlay' rolls out in AKS-RP
-			if (invoker.ipamMode != util.V4Overlay) && (invoker.ipamMode != util.DualStackOverlay) && (invoker.ipamMode != util.Overlay) {
-				return IPAMAddResult{}, errors.Wrap(errInvalidArgs, "%w: Gateway address "+info.ncGatewayIPAddress+" from response is invalid")
+			macAddress, err := net.ParseMAC(info.macAddress)
+			if err != nil {
+				return IPAMAddResult{}, errors.Wrap(err, "Invalid mac address")
 			}
 
-			if net.ParseIP(info.podIPAddress).To4() != nil { //nolint:gocritic
-				ncgw, err = getOverlayGateway(ncIPNet)
+			isDefaultInterfaceSet = isDefaultInterfaceSet || info.isDefaultInterface
+			result := CNIResult{
+				ipResult: &cniTypesCurr.Result{
+					IPs: []*cniTypesCurr.IPConfig{
+						{
+							Address: net.IPNet{
+								IP:   ip,
+								Mask: ipnet.Mask,
+							},
+						},
+					},
+					Interfaces: []*cniTypesCurr.Interface{
+						{
+							Mac: info.macAddress,
+						},
+					},
+				},
+				addressType:        cns.Secondary,
+				macAddress:         macAddress,
+				isDefaultInterface: info.isDefaultInterface,
+			}
+
+			for _, route := range info.routes {
+				_, dst, err := net.ParseCIDR(route.IPAddress)
```

Comment:
should we be using netip package instead of net?

Reply:
we'll need ParseCIDR since we expect input to be in CIDR notation